File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sortedArrayToBST(nums: number[]): TreeNode | null {
  const N = nums.length;
  
  if(N <= 0)
   return null;
  
  const half = N >>> 1;
  return new TreeNode(nums[half], sortedArrayToBST(nums.slice(0, half)), sortedArrayToBST(nums.slice(1 + half)));
};

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

	public TreeNode sortedArrayToBST(int[] nums) {
		return this.sortedArrayToBST(nums, 0, nums.length - 1);
	}

	public TreeNode sortedArrayToBST(int[] nums, int low, int hi) {
		if (low > hi)
			return null;

		int mid = low + (hi - low) / 2;
		TreeNode node = new TreeNode(nums[mid]);
		node.left = this.sortedArrayToBST(nums, low, mid - 1);
		node.right = this.sortedArrayToBST(nums, mid + 1, hi);

		return node;
	}    
}